File: Lexer/ID.py
from AFD import AFD
from Token import Token

class ID(AFD):
    def evaluate(self, code, line, column):
        code_iter = iter(code)
        current_char = next(code_iter, None)  # Obtém o próximo caractere da sequência
        if current_char is not None and current_char.isalpha():  # Verifica se o caractere é uma letra
            identifier = self.read_id(code_iter, current_char)  # Lê o identificador
            if self.end_id(code_iter):  # Verifica se o identificador terminou
                return Token("ID", identifier, line, column)
        return None

    def read_id(self, code_iter, current_char):
        identifier = ""
        while current_char is not None and current_char.isalnum():  # Enquanto o caractere for uma letra ou um número
            identifier += current_char
            current_char = next(code_iter, None)  # Obtém o próximo caractere
        return identifier

    def end_id(self, code_iter):
        # Verifica se o próximo caractere não é alfanumérico
        return next(code_iter, None) is None

    def skip_white_space(self, code):  # Adiciona esta função
        while code.current() in (' ', '\n', '\r', '\t'):
            code.next()

File: src/alvaro/Lexico/ID.java
package Lexico;

import java.text.CharacterIterator;

public class ID extends AFD {
    @Override
    public Token evaluate(CharacterIterator code, int line, int column) {
        if (Character.isLetter(code.current())) {
            String number = readID(code);

            if (endID(code)) {
                return new Token("ID", number, line, column);
            }
        }

        return null;
    }

    private String readID(CharacterIterator code) {
        String id = "";

        while (Character.isLetter(code.current())) {
            id += code.current();
            code.next();
        }

        return id;
    }

    private boolean endID(CharacterIterator code) {
        return code.current() == ' '
                || code.current() == '+'
                || code.current() == '-'
                || code.current() == '*'
                || code.current() == '('
                || code.current() == ')'
                || code.current() == '{'
                || code.current() == '}'
                || code.current() == '['
                || code.current() == ']'
                || code.current() == ';'
                || code.current() == '\r'
                || code.current() == '\n'
                || code.current() == CharacterIterator.DONE;
    }
}
